### PR TITLE
Add new hook `HookLoadFileExtended` that allows plugins to supply Zeek script code to parse.

### LINKED ISSUE #1757: Extend the `LoadFile` plugin hook so that it can provide a script's content
The current `LoadFile` hook can only either (1) fully take care of processing loading a script, or leave it all to Zeek for normal processing. It would be useful if the hook could do it "half-way": provide the *content* of the script to be loaded, but then leave parsing that content to Zeekr. That way, a plugin could pull script code from different places than just the standard path on disk.

### PR DESCRIPTION
**Pull request by rsmmr**
The new hook works similar to the existing `HookLoadFile` but,
additionally, allows the plugin to return a string that contains the
code to be used for the file being loaded. If the plugin does so, the
content of any actual file on disk will be ignored (in fact, there
doesn't even need to be a file on disk in that case). This works for
both Zeek scripts and signatures.

There's a new test that covers the new functionality, testing loading
both scripts and signatures from memory. I also manually tested that the
debugger integration works, but I don't see much of a way to add a
regression test for that part.

We keep the existing hook as well for backwards compatibility. We could
decide to deprecate it, but not sure that buys us much, so left that
out.

Closes #1757.

### CODE REVIEW

**Comment by rsmmr** [comment]
This is ready for review.

**Review by 0xxon — APPROVED** [review]
Looks good to me.

Just one random question - does this give reasonable-ish error messages if the passed in script/signature has errors and cannot be parsed by Zeek?

**Comment by rsmmr** [comment]
> Just one random question - does this give reasonable-ish error messages if the passed in script/signature has errors and cannot be parsed by Zeek?

Errors should be just as if it was loaded normally from disk. Do you think that could be confusing? Not sure how we'd track the information that content came from a plugin, would need to get into the `Location` objects I guess. Plus, that way scripts from a plugin would then be distinguishable from normal scripts, which may actually not be desired depending on use case (because maybe they *are* the normal scripts, just loaded from a different place).

File: src/PolicyFile.cc [hunk 1]
```diff
@@ -72,49 +72,60 @@ int how_many_lines_in(const char* policy_filename)
...
 		}
+	else
+		{
+		FILE* f = fopen(policy_filename, "r");
 
-	pf->lmtime = st.st_mtime;
-	off_t size = st.st_size;
+		if ( ! f )
+			{
+			debug_msg("No such policy file: %s.\n", policy_filename);
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply src/PolicyFile.cc:100]: Just change to `Could not open ...` ?

[the remark this answers is not in the document]
Comment (0xxon) [reply src/PolicyFile.cc:100]: Sounds good to me.

**Comment by 0xxon** [comment]
> Errors should be just as if it was loaded normally from disk. Do you think that could be confusing? Not sure how we'd track the information that content came from a plugin, would need to get into the `Location` objects I guess. Plus, that way scripts from a plugin would then be distinguishable from normal scripts, which may actually not be desired depending on use case (because maybe they _are_ the normal scripts, just loaded from a different place).

Well, I can see that it could be confusing when the error message shows something different than the file content. However - I also don't see a great easy way around that, so that is fine. Thank you for satisfying my curiosity.

**Pull request merged**